content/courses/advanced-ada/chapters/limited_types.rst

```diff
@@ -425,6 +431,566 @@ required.
+as :ada:`private limited`. However, :ada:`Rec_Limited_Full` type is limited in
+its full view, while :ada:`Rec_Nonlimited_Full` is nonlimited. As expected,
+the compiler complains about the :ada:`To := From` assignment in the
+:ada:`Copy` procedure for the :ada:`Rec_Limited_Full` type because its full
+view is limited (so no assignment is possible). Of course, in the case of the
+objects of :ada:`Rec_Nonlimited_Full` type, this assignment is perfectly fine.
+
+
+Tagged limited private types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For tagged private types, the partial and full views must match: if a tagged
+type is limited in the partial view, it must be limited in the full view. For
+example:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Limited_Types.Tagged_Limited_Private_Types
+
+    package Simple_Recs is
+
+       type Rec is tagged limited private;
+
+    private
+
+       type Rec is tagged limited record
+          I : Integer;
+       end record;
+
+    end Simple_Recs;
+
+Here, the tagged :ada:`Rec` type is limited both in its partial and full views.
+Any mismatch in one of the views triggers a compilation error. (As an
+exercise, you may remove any of the :ada:`limited` keywords from the code
+example and try to compile it.)
+
+
+Deriving from limited types
+---------------------------
+
+In this section, we discuss the implications of deriving from limited types.
+As usual, let's start with a simple example:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Limited_Types.Derived_Limited_Type
+
+    package Simple_Recs is
+
+       type Rec is limited null record;
+
+       type Rec_Derived is new Rec;
+
+    end Simple_Recs;
+
+In this example, the :ada:`Rec_Derived` type is derived from the :ada:`Rec`
+type. Note that the :ada:`Rec_Derived` type is limited because its ancestor is
+limited, even though the :ada:`limited` keyword doesn't show up in the
+declaration of the :ada:`Rec_Derived` type. Note that we could have actually
+used the :ada:`limited` keyword here:
+
+.. code-block:: ada
+
+       type Rec_Derived is limited new Rec;
+
+Therefore, we cannot use the assignment operator for objects of
+:ada:`Rec_Derived` type:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Limited_Types.Derived_Limited_Type
+    :class: ada-expect-compile-error
+
+    with Simple_Recs; use Simple_Recs;
+
+    procedure Test_Limitedness is
+       Dummy_1, Dummy_2 : Rec_Derived;
+    begin
+       Dummy_2 := Dummy_1;
+    end Test_Limitedness;
+
+Note that we cannot derive a limited type from a nonlimited ancestor:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Limited_Types.Derived_Limited_Type_Nonlimited_Ancestor
+    :class: ada-expect-compile-error
+
+    package Simple_Recs is
+
+       type Rec is null record;
+
+       type Rec_Derived is limited new Rec;
+
+    end Simple_Recs;
+
+As expected, the compiler indicates that :ada:`Rec` should be of limited type.
+
```

Comment:
As a result, all the types in a derivation class are the same -- either limited or not. That will be especially important with dynamic dispatching via tagged types.

Reply:
Good point, thanks! I've just added this to the text.